##  Evaluación comparativa

Grupo: 1

Estudiantes:

-Constanza Olivos Fernandez

-Javier Nanco Becerra

-Nicolás Pozo Villagrán

Fecha: 20-09-2025

Version: 1.4

# Objetivos del notebook

En esta sección se realiza una comparación integral del desempeño de todos los modelos entrenados (clásicos, modernos y el MLP optimizado) utilizando el conjunto de prueba. Para ello, se calculan y consolidan las métricas MAE, RMSE, R² y MAPE, las cuales permiten evaluar tanto la precisión como la estabilidad de cada modelo.

El proceso incluye:

La lectura de los archivos de resultados individuales generados en etapas anteriores.

La unificación de todas las métricas en un único DataFrame.

La clasificación de los modelos en dos grupos: Anteriores (modelos clásicos y boosting) y Optimizados (modelos modernos y MLP optimizado).

La ordenación de los resultados por R² descendente, lo que permite identificar los modelos con mejor capacidad predictiva.

La exportación de la tabla final a un archivo CSV para su posterior análisis y visualización en informes o herramientas externas.

De este modo, se obtiene una tabla comparativa completa que facilita la identificación de los modelos más robustos y consistentes, brindando evidencia objetiva para la selección del mejor enfoque en la predicción del rendimiento del trigo.

In [12]:
# ===========================
# Comparación final distinguida
# ===========================

import pandas as pd
import glob

# ===========================
# 1. Leer todos los CSV de resultados
# ===========================
archivos_resultados = glob.glob("csv/resultados_*.csv")  # Ajusta según tus archivos

lista_df = []
for archivo in archivos_resultados:
    df = pd.read_csv(archivo)
    
    # Agregar columna para distinguir optimizados
    if "mlp" in archivo.lower() or "modernos" in archivo.lower():
        df["Estado"] = "Optimizado"
    else:
        df["Estado"] = "Anterior"
    
    lista_df.append(df)

# ===========================
# 2. Combinar todos los resultados
# ===========================
df_comparativo = pd.concat(lista_df, ignore_index=True)

# ===========================
# 3. Ordenar por R² descendente dentro de cada grupo
# ===========================
df_comparativo = df_comparativo.sort_values(by=["Estado", "R²"], ascending=[False, False])

# ===========================
# 4. Separar por grupos para ver más claro
# ===========================
df_optimizado = df_comparativo[df_comparativo["Estado"] == "Optimizado"]
df_anterior = df_comparativo[df_comparativo["Estado"] == "Anterior"]

print("✅ Modelos Optimizado / Mejorado:")
print(df_comparativo)


import os

# ===========================
# 5. Guardar CSV final distinguido en la carpeta 'csv'
# ===========================
output_path = os.path.join('csv', 'comparacion_modelos_trigo_distinguido.csv')
df_comparativo.to_csv(output_path, index=False)
print("✅ Tabla comparativa guardada en 'csv/comparacion_modelos_trigo_distinguido.csv'")



✅ Modelos Optimizado / Mejorado:
              Modelo        R²     R²_CV        RMSE         MAE  MAPE (%)  \
10          CatBoost  0.830510       NaN  304.688421  239.515335  4.881455   
9           LightGBM  0.826762       NaN  308.039440  242.200684  4.937405   
8            XGBoost  0.825747       NaN  308.939874  243.050112  4.957638   
0                MLP  0.794230  0.732465  335.718683  263.637716  5.384025   
5      Random Forest  0.819684       NaN  314.268494  247.256763  5.041656   
7                KNN  0.760639       NaN  362.085107  286.914409  5.911103   
1   Regresión Lineal  0.701081       NaN  404.632495  315.828222  6.565522   
3              Lasso  0.701075       NaN  404.636908  315.850907  6.565973   
2              Ridge  0.701032       NaN  404.666045  315.979182  6.568551   
6                SVR  0.643299       NaN  442.014200  352.208376  7.435710   
4              Árbol  0.639440       NaN  444.398956  351.665011  7.173763   

        Estado  
10  Optimizad

La evaluación comparativa evidencia que los modelos modernos optimizados (CatBoost, LightGBM y XGBoost) superan consistentemente a los modelos clásicos en términos de precisión y error. CatBoost se posiciona como el mejor modelo, alcanzando un R² = 0.83 y el menor error absoluto (MAE ≈ 239.5), lo que refleja alta capacidad predictiva y generalización confiable.

El MLP optimizado también mostró un buen desempeño (R² = 0.79, MAPE ≈ 5.38%), quedando por debajo de los modelos de boosting, pero por encima de varios algoritmos clásicos. En contraste, los métodos lineales (Regresión Lineal, Ridge y Lasso) y los modelos de menor complejidad (Árbol de decisión, SVR) presentaron limitaciones claras, con R² ≤ 0.70 y mayores niveles de error.

En conjunto, los resultados sugieren que los enfoques de boosting (CatBoost, LightGBM y XGBoost) son los más adecuados para predecir el rendimiento del trigo, entregando estimaciones más estables y precisas, mientras que los modelos clásicos solo resultan útiles como referencia base o para escenarios con menor complejidad.